udaykalra/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,133 @@
+        }
+      }
+    
+
+    // Eliminate invalid time intervals by duration.
+    for (TimeRange index : availableRanges) {
+      if (index.duration() < requestDuration) {
+        availableRanges.remove(index);
+      }
+    }
+
+    // Convert to list and sort.
+    List outList = new ArrayList(availableRanges);
+    Collections.sort(outList, TimeRange.ORDER_BY_START);
+    return outList;
   }
+
+  /**
+   * Checks whether or not a given time range overlaps with a given busy range and cuts the event
+   * shorter in a collection if needed.
+   */
+  private void overlapCut(TimeRange givenRange, TimeRange busyRange, Collection<TimeRange> whereCut){
+      // Obtain output and busy time range data.
+        int givenStart = givenRange.start();
+        int givenEnd = givenRange.end();
+
+        int busyStart = busyRange.start();
+        int busyEnd = busyRange.end();
+
+        if (givenRange.overlaps(busyRange)) {
+          whereCut.remove(givenRange);
+          /*
+           * Case 1: Busy meeting at start of free interval and ends sooner.
+           * Reduce length of free interval.
+           */
+          if ((givenStart == busyStart) && (givenEnd > busyEnd)) {
+            TimeRange replaceTime = TimeRange.fromStartDuration(busyEnd, givenEnd - busyEnd);
+            whereCut.add(replaceTime);
+          }
+          /*
+           * Case 2: Busy meeting starts during free interval and ends at same time.
+           * Reduce length of free interval.
+           */
+          else if ((givenEnd == busyEnd) && (givenStart < busyStart)) {
+            TimeRange replaceTime = TimeRange.fromStartDuration(givenStart, busyStart - givenStart);
+            whereCut.add(replaceTime);
+          }
+          /*
+           * Case 3: Busy meeting starts during free interval and ends after.
+           * Reduce length of free interval.
+           */
+          else if ((givenStart < busyStart) && (givenEnd < busyEnd)) {
+            TimeRange replaceTime = TimeRange.fromStartDuration(givenStart, busyStart - givenStart);
+            whereCut.add(replaceTime);
+          }
+          /*
+           * Case 4: Busy meeting is completely overlapped by free interval.
+           * Chop out of free interval.
+           */
+          else if ((givenStart < busyStart) && (givenEnd > busyEnd)) {
+            TimeRange replaceTimeA =
+                TimeRange.fromStartDuration(givenStart, busyStart - givenStart);
+            TimeRange replaceTimeB = TimeRange.fromStartDuration(busyEnd, givenEnd - busyEnd);
+            whereCut.add(replaceTimeA);
+            whereCut.add(replaceTimeB);
+          }
+        }
+        if (busyRange.overlaps(givenRange)) {
+          whereCut.remove(givenRange);
+          /*
+           * Case 5: Busy meeting starts before free interval and ends during.
+           * Reduce length of free interval.
+           */
+          if ((givenStart > busyStart) && (givenEnd > busyEnd)) {
```

Comment:
// Case 5: `possiblyOverlappingTimeRange` starts before this TimeRange and ends during this TimeRange.
// The start of TimeRange needs to be trimmed.

Suggest assigning the value of this statement to a local variable to make it easier to distinguish between the various if-conditions.

Reply:
Done